routing/index_graph.cpp

```diff
@@ -13,27 +16,520 @@ IndexGraph::IndexGraph(unique_ptr<GeometryLoader> loader, shared_ptr<EdgeEstimat
+    {
+      if (e.GetTarget() == jointId)
+        oneStepAside.push_back(jointId);
+    }
+  });
+}
+
+bool IndexGraph::GetIngoingAndOutgoingEdges(Joint::Id centerId, bool graphWithoutRestrictions,
+                                            vector<JointEdge> & ingoingEdges,
+                                            vector<JointEdge> & outgoingEdges)
+{
+  ingoingEdges.clear();
+  outgoingEdges.clear();
+  GetEdgeList(centerId, false /* isOutgoing */, graphWithoutRestrictions, ingoingEdges);
+  if (ingoingEdges.empty())
+    return false;
+
+  GetEdgeList(centerId, true /* isOutgoing */, graphWithoutRestrictions, outgoingEdges);
+  if (outgoingEdges.empty())
+    return false;
+  return true;
+}
+
+bool IndexGraph::ApplyRestrictionPrepareData(RestrictionPoint const & restrictionPoint,
+                                             RestrictionInfo & restrictionInfo)
+{
+  vector<JointEdge> ingoingEdges;
+  vector<JointEdge> outgoingEdges;
+
+  GetEdgeList(restrictionPoint.m_centerId, false /* isOutgoing */, true /* graphWithoutRestrictions */, ingoingEdges);
+  vector<Joint::Id> fromOneStepAside;
+  FindOneStepAsideRoadPoint(restrictionPoint.m_from, restrictionPoint.m_centerId, ingoingEdges,
+                            fromOneStepAside);
+  if (fromOneStepAside.empty())
+    return false;
+
+  GetEdgeList(restrictionPoint.m_centerId, true /* isOutgoing */, true /* graphWithoutRestrictions */, outgoingEdges);
+  vector<Joint::Id> toOneStepAside;
+  FindOneStepAsideRoadPoint(restrictionPoint.m_to, restrictionPoint.m_centerId, outgoingEdges,
+                            toOneStepAside);
+  if (toOneStepAside.empty())
+    return false;
+
+  restrictionInfo.m_center = restrictionPoint.m_centerId;
+  restrictionInfo.m_from = fromOneStepAside.back();
+  restrictionInfo.m_to = toOneStepAside.back();
+  restrictionInfo.m_fromFeatureId = restrictionPoint.m_from.GetFeatureId();
+  restrictionInfo.m_toFeatureId = restrictionPoint.m_to.GetFeatureId();
+  return true;
+}
+
+void IndexGraph::ApplyRestrictionNoRealFeatures(RestrictionPoint const & restrictionPoint)
+{
+  ApplyRestrictionRealFeatures(restrictionPoint, [&] (RestrictionInfo const & restrictionInfo) {
+    ApplyRestrictionNo(restrictionInfo);
+  });
+}
+
+void IndexGraph::ApplyRestrictionRealFeatures(RestrictionPoint const & restrictionPoint,
+                                              function<void(RestrictionInfo const &)> && f)
+{
+  RestrictionInfo restrictionInfo;
+  if (!ApplyRestrictionPrepareData(restrictionPoint, restrictionInfo))
+    return;
+
+  // Note. It's necessary to collect edges ingoing to the restriction and
+  // outgoing from the restriontion at first and then to apply |f|
+  // because |f| edits the graph and excecuting |f| affect the behavior of
+  // ForEachNonBlockedEdgeMappingNode().
+  auto const edges = restrictionInfo.Edges();
+  vector<DirectedEdge> ingoingRestEdges;
+  ForEachNonBlockedEdgeMappingNode(edges.first, [&](DirectedEdge const & ingoing){
```

Comment:
It does not work this way.  |f| is ApplyRestrictionNo() and ApplyRestrictionOnly(). So when f is called |m_blockedEdges| and |m_edgeMapping| is changed. This changing affect ForEachNonBlockedEdgeMappingNode(). But before adding a restriction we need to get all duplicates of ingoing and outgoing edge.